File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        
        // 유저 목록 저장
        Map<String, Set<String>> reports = new HashMap<>();
        
        // 신고 횟수 저장
        Map<String, Integer> reportCounts = new HashMap<>();
        
        // 초기화
        for (String id : id_list) {
            reports.put(id, new HashSet<>());
            reportCounts.put(id, 0);
        }
        
        // 신고 처리
        for (String rep : report) {
            String[] parts = rep.split(" ");
            String reporter = parts[0];
            String reported = parts[1];
            
            // 중복 처리
            if (!reports.get(reporter).contains(reported)) {
                reports.get(reporter).add(reported);
                reportCounts.put(reported, reportCounts.get(reported) + 1);
            }
        }
        
        // 정지 유저 목록 확인
        Set<String> suspendedUsers = new HashSet<>();
        for (String id : id_list) {
            if (reportCounts.get(id) >= k) {
                suspendedUsers.add(id);
            }
        }
        
        // 결과 메일 발송 횟수
        int[] answer = new int[id_list.length];
        for (int i = 0; i < id_list.length; i++) {
            String user = id_list[i];
            int mailCount = 0;
            for (String reported : reports.get(user)) {
                if (suspendedUsers.contains(reported)) {
                    mailCount++;
                }
            }
            answer[i] = mailCount;
        }
        
        return answer;
    }
}

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.py
def solution(id_list, report, k):
    
    # 1. 신고 기록 정리 (중복 제거)
    report_dict = {user: set() for user in id_list}
    for rep in report:
        user, reported = rep.split()
        report_dict[user].add(reported)
    
    # 2. 신고 횟수 계산
    reported_count = {user: 0 for user in id_list}
    for reports in report_dict.values():
        for reported in reports:
            reported_count[reported] += 1
    
    # 3. 정지된 유저 확인
    suspended_users = {user for user, count in reported_count.items() if count >= k}
    
    # 4. 각 유저별 결과 메일 수 계산
    result = []
    for user in id_list:
        result.append(len(report_dict[user] & suspended_users))
    
    return result
